### ISSUE TITLE
Add support for double-precision floating-point 

### ISSUE THREAD

**Issue by alexsavulescu** [BlueBrain/MorphIO#162]
We would need to have the possibility of using Morphio with doubles, as well as its python bindings.  For example, NRN is using doubles for points and their computation. 

After discussing with @wizmer and @ferdonline, we could consider one of the following: 
- CMAKE option to enable double support
- just use double as default

TBD

**Comment by mgeplf** [comment]
In the original design, space was a big concern: the viz team brought up the fact they keep many morphs in memory at once, and since the source data is of low precision (float in hdf5, for Neurolucida ASCII, the points only have 3 decimal points - plus these positions are not very precise in the first place), that doubling the space requirement wouldn't bring much to the table.

**Comment by alexsavulescu** [comment]
What I'm working on is a way to use MorphIo for morphology loading in NRN. Since the library is parsing floats and then passing back numpy ndarray of floats, this impacts the soma points calculations so we end up with different results. True,these differences might seem insignificant as they slightly differ (i.e. -0.576182 vs -0.576181) but they will have an impact in the overall simulation run, leading to different number of spikes and results a bit off from what you would get with NRN morphology loading. 

Would you then agree to have a CMAKE option to enable double support?

**Comment by mgeplf** [comment]
> Would you then agree to have a CMAKE option to enable double support?

I'm not against it, I'm just not sure it will solve your problem: like I said, h5 files store floats internally.  The ASCII files might give the same results.

Keeping the (already complex) library complexity down is also nice, so the changes would hopefully not be too invasive.

**Comment by alexsavulescu** [comment]
I see your point. Only worked with ASCII so far, I have to check h5 out. 

The changes would not be too invasive :)

**Comment by wizmer** [comment]
Let's summarize.

**1) Keeping float in MorphIO**

Plus: 
- Size stay minimal
- Less complexity

Minus:
NRN loader and MorphIO loader differ for ASC files


**2) CMake flag:**

Plus:
NRN loader and MorphIO loader can be made equivalent for ASC files

Minus:
- More complexity
- MorphIO will now have multiples outcome because of the flag.

Question:
Should the flag propagate to the python bindings ?

**3) Move MorphIO to double:**

Plus:
NRN and MorphIO now behaves exactly the same way.

Minus:
Double the size of the morphology object.

Note:
From what I have seen, Viz team would probably discard the morphology after fetching the points and the relevant information they need. They could probably cast back to float there if needed.

===============================================================
I'd say it could be nice to have the double flag for testing purposes. IE. to check that both loaders are actually giving the same results and that everything is working fine.
Once this is done, I think reverting to floats would make sense. Yes, that would means that NRN outcome will be different, but we'll know by then that this is a floating point issue. Even if spikes positions move, the statistical properties of the simulation should be maintained. Let's not forget 
we are talking about going from `float` to `double` for experimental coordinates that have been grossly measured. I would argue that NRN choice of using double for something inherently imprecise is debatable.

**Comment by alexsavulescu** [comment]
@ferdonline we should have a discussion on this in order to decide on the best way to move forward.

**Comment by mgeplf** [comment]
One method to reduce the problems w/ the float32 h5 representation, and the ASC files is to use the following the writer:

```diff
@@ -138,9 +145,12 @@ static void _write_asc_points(std::ofstream& myfile,
                               const std::vector<float>& diameters,
                               size_t indentLevel) {
     for (unsigned int i = 0; i < points.size(); ++i) {
-        myfile << std::string(indentLevel, ' ') << '(' << std::to_string(points[i][0]) << ' '
-               << std::to_string(points[i][1]) << ' ' << std::to_string(points[i][2]) << ' '
-               << std::to_string(diameters[i]) << ")\n";
+        myfile << std::fixed << std::setprecision(FLOAT_PRECISION_PRINT)
+            << std::string(indentLevel, ' ') << '('
+            << points[i][0] << ' '
+            << points[i][1] << ' '
+            << points[i][2] << ' '
+            << diameters[i] << ")\n";
     }
 }
```
With `FLOAT_PRECISION_PRINT = 9`

9 is chosen so that the each float32 is unique, and then the double will be the same.

In theory, if you're transforming the soma contour -> cylinder stack, you'll need to upconvert the points to `double` first.

Does that work for you @alexsavulescu

**Comment by alexsavulescu** [comment]
@mgeplf I was just using MorphIO through Python to read an ASC file from the circuit and comparing the values to regular Import_3D. By upconverting float32s to double the precision issue was still there. The only way of getting the same results was to modify MorphIO and read doubles from the file(using stod here: https://github.com/BlueBrain/MorphIO/blob/4f8b7b17bf014f6c4669538f22fa30de3f3c1460/src/readers/morphologyASC.cpp#L72 )  and then pass an ndarray of doubles.

I didn't spend much time on testing h5 loading through MorphIO, but in the circuit I was working on the values were stored as float64: 
```
DF5 "dend-mpg141217_A_idB_axon-mpg141017_a1-2_idC_-_Clone_7.h5" {
GROUP "/" {
   DATASET "points" {
      DATATYPE  H5T_IEEE_F64LE
```

**Comment by mgeplf** [comment]
True; for the 'random' ascii files that are out there, we'll have that issue. (And also h5's that don't adhere to the standard) 

However, by upping the number of digits output when doing h5 -> ascii conversion (and swc), I can get the same numbers in neuron when loading files (except, of course, the soma), so we can make sure we're loading things right, and then it's a matter of deciding whether the simulation outputs that differ due to precision matter.

IE: at some point in the future, things will be stable...after much work.

**Comment by alexsavulescu** [comment]
So the idea would be to re-generate the ASCII files? Though I wonder, why do we store things as float64 and not float32 in H5? Or is it just another remnant ? 

Yes, the main issue to deal with will be the soma calculations because neuron uses doubles.

**Comment by mgeplf** [comment]
> So the idea would be to re-generate the ASCII files?

That's actually how it happens: h5 files are converted to ASC files during the morphology release process.

>Though I wonder, why do we store things as float64 and not float32 in H5? Or is it just another remnant ?

It was a concious decision to reduce storage space: the incoming data is not precise, so it doesn't seem worthwhile to keep much precision.

> Yes, the main issue to deal with will be the soma calculations because neuron uses doubles.

Yeah; if we keep enough digits to ensure float32 uniqueness, then as long as all calculations are done w/ the same precision, we're ok.

**Comment by wizmer** [comment]
> It was a concious decision to reduce storage space: the incoming data is not precise, so it doesn't seem worthwhile to keep much precision.
Well, what @alexsavulescu saw is that in the H5, things ARE stored in 64 bits ! I don't know why, probably another remnant yes...

> Yeah; if we keep enough digits to ensure float32 uniqueness, then as long as all calculations are done w/ the same precision, we're ok.
I'll make the PR to change the precsion then.

**Comment by wizmer** [comment]
@mgeplf I think we should start by fixing these non-conform H5 files that use float64. I could add a CLI in morph-tool to fix them ?

**Comment by wizmer** [comment]
Summarizing the short discussion I had with @alexsavulescu , since he already locally have the requested feature, he will work with that for the time being. On my side I will:

- sanitize the H5 files that store `double` instead of `float`
- Increase `setprecision` when writing ASC and SWC files.

I'm closing this PR but if the feature is still needed at the end of @alexsavulescu 's work, feel free to reopen it.

**Comment by alexsavulescu** [comment]
@wizmer I'd like to tackle this. During our last meeting we said to start off with adding a CMAKE option to support doubles and then see from there.

**Comment by wizmer** [comment]
Yes, sure. You probably already have everything you need for that ?

**Comment by alexsavulescu** [comment]
Normally yes. Did have a bit of code somewhere, need to dig that out. Will let you know if I need some help.
